mission_control/src/mission.cpp

```diff
@@ -36,46 +35,175 @@
+{
+  // TODO(hidmic): fetch pitch angle from blackboard
+  static constexpr char text[] = R"(
+<root main_tree_to_execute="Go to the surface" >
+    <BehaviorTree ID="Go to the surface">
+       <AttitudeServoBehavior name="Command thruster and fins"
+           roll="0.0" pitch="-20.0" yaw="0.0" speed_knots="0.0"/>
+    </BehaviorTree>
+</root>
+)";
+
+  MissionBehaviorTreeFactory& factory =
+      MissionBehaviorTreeFactory::instance();
+  abort_behavior_tree_ = factory.createTreeFromText(text);
+}
+
+std::unique_ptr<Mission> Mission::fromFile(const std::string& path)
+{
+  if (access(path.c_str(), F_OK) < -1)
+  {
+    ROS_ERROR_STREAM("Cannot read mission definition from " << path);
     return nullptr;
+  }
+
+  MissionBehaviorTreeFactory& factory = MissionBehaviorTreeFactory::instance();
+  return std::unique_ptr<Mission>(new Mission(factory.createTreeFromFile(path)));
 }
 
-void Mission::stop()
+const std::string& Mission::description() const
 {
-  tree_.haltTree();
-  behaviorStatus_ = BT::NodeStatus::IDLE;
+  return main_behavior_tree_.rootNode()->name();
 }
 
-BT::NodeStatus Mission::getStatus() { return behaviorStatus_; }
+bool Mission::active() const
+{
+  return (status_ != Mission::Status::READY &&
+          status_ != Mission::Status::PREEMPTED &&
+          status_ != Mission::Status::COMPLETED &&
+          status_ != Mission::Status::ABORTED);
+}
 
-std::string Mission::getCurrentMissionDescription() { return description_; }
+Mission& Mission::start()
+{
+  switch (status_)
+  {
+    case Mission::Status::COMPLETED:
+    case Mission::Status::ABORTED:
+    case Mission::Status::READY:
+      status_ = Mission::Status::PENDING;
+      break;
+    default:
+      break;
+  }
+  return *this;
+}
 
-void Mission::Continue()
+Mission& Mission::resume()
 {
-  if (behaviorStatus_ == BT::NodeStatus::IDLE || behaviorStatus_ == BT::NodeStatus::RUNNING)
+  switch (status_)
   {
-    behaviorStatus_ = tree_.tickRoot();
+    case Mission::Status::PENDING:
+    case Mission::Status::EXECUTING:
+      switch (main_behavior_tree_.tickRoot())
+      {
+        case BT::NodeStatus::RUNNING:
+          status_ = Mission::Status::EXECUTING;
+          break;
+        case BT::NodeStatus::SUCCESS:
+          status_ = Mission::Status::COMPLETED;
+          break;
+        case BT::NodeStatus::FAILURE:
+          status_ = Mission::Status::ABORTING;
+          break;
+      }
+      break;
+    case Mission::Status::ABORTING:
+      switch (abort_behavior_tree_.tickRoot())
+      {
+        case BT::NodeStatus::RUNNING:
+          status_ = Mission::Status::ABORTING;
+          break;
+        case BT::NodeStatus::SUCCESS:
+        case BT::NodeStatus::FAILURE:
+          status_ = Mission::Status::ABORTED;
+          break;
+      }
+      break;
+    default:
+      break;
   }
+  return *this;
 }
+
+Mission& Mission::preempt()
+{
+  switch (status_)
+  {
+    case Mission::Status::PENDING:
+      status_ = Mission::Status::PREEMPTED;
```

Comment:
nit: status are defined as PREEMPTED in all the cases.

Reply:
Not in all cases, but in three (3) states. And for each one of those cases, something special gets done.

I find this easier to read, despite the duplication. What would you propose?

Reply:
+1 for the sake of readability